File: algoritmos/problema 1 Terminal Inteligente/algoritmoFuerzaBruta.py
import sys

COST_ADVANCE = 1
COST_DELETE = 2
COST_INSERT = 2
COST_REPLACE = 3
COST_KILL = 1


def algoritmoFuerzaBruta(cadena_original, cadena_objetivo):
    operaciones = []
    cadena_formada = ""
    conteo_operaciones = {"advance": 0, "delete": 0, "insert": 0, "replace": 0, "kill": 0}
    costo_total, cadena_formada = algoritmoFuerzaBrutaRecursiva(cadena_original, cadena_objetivo, 0, 0, 0, operaciones, conteo_operaciones, cadena_formada)
    return costo_total, operaciones, conteo_operaciones, cadena_formada

def algoritmoFuerzaBrutaRecursiva(cadena_original, cadena_objetivo, i, j, costo_total, operaciones, conteo_operaciones, cadena_formada):

    if i == len(cadena_original) and j == len(cadena_objetivo):
        return costo_total, cadena_formada
    
    if i < len(cadena_original) and j < len(cadena_objetivo) and cadena_original[i] == cadena_objetivo[j]:
        operaciones.append(f"advance at position {i}")
        conteo_operaciones["advance"] += 1
        cadena_formada += cadena_original[i]  
        return algoritmoFuerzaBrutaRecursiva(cadena_original, cadena_objetivo, i+1, j+1, costo_total + COST_ADVANCE, operaciones, conteo_operaciones, cadena_formada)

    if i < len(cadena_original) and j < len(cadena_objetivo):
        operaciones_reemplazo = operaciones[:]
        operaciones_reemplazo.append(f"replace {cadena_original[i]} with {cadena_objetivo[j]} at position {i}")
        conteo_operaciones_reemplazo = conteo_operaciones.copy()
        conteo_operaciones_reemplazo["replace"] += 1
        cadena_reemplazo = cadena_formada + cadena_objetivo[j]  
        reemplazo, cadena_reemplazo = algoritmoFuerzaBrutaRecursiva(cadena_original, cadena_objetivo, i+1, j+1, costo_total + COST_REPLACE, operaciones_reemplazo, conteo_operaciones_reemplazo, cadena_reemplazo)
    else:
        reemplazo = sys.maxsize

    if j < len(cadena_objetivo):
        operaciones_insercion = operaciones[:]
        operaciones_insercion.append(f"insert {cadena_objetivo[j]} at position {i}")
        conteo_operaciones_insercion = conteo_operaciones.copy()
        conteo_operaciones_insercion["insert"] += 1
        cadena_insercion = cadena_formada + cadena_objetivo[j] 
        insercion, cadena_insercion = algoritmoFuerzaBrutaRecursiva(cadena_original, cadena_objetivo, i, j+1, costo_total + COST_INSERT, operaciones_insercion, conteo_operaciones_insercion, cadena_insercion)
    else:
        insercion = sys.maxsize

    if i < len(cadena_original):
        operaciones_borrado = operaciones[:]
        operaciones_borrado.append(f"delete {cadena_original[i]} at position {i}")
        conteo_operaciones_borrado = conteo_operaciones.copy()
        conteo_operaciones_borrado["delete"] += 1
        cadena_borrado = cadena_formada  
        borrado, cadena_borrado = algoritmoFuerzaBrutaRecursiva(cadena_original, cadena_objetivo, i+1, j, costo_total + COST_DELETE, operaciones_borrado, conteo_operaciones_borrado, cadena_borrado)
    else:
        borrado = sys.maxsize

    if i < len(cadena_original) and j == len(cadena_objetivo):
        operaciones_kill = operaciones[:]
        operaciones_kill.append(f"kill from position {i}")
        conteo_operaciones_kill = conteo_operaciones.copy()
        conteo_operaciones_kill["kill"] += 1
        cadena_kill = cadena_formada  
        kill = costo_total + COST_KILL
    else:
        kill = sys.maxsize

    min_costo = min(reemplazo, insercion, borrado, kill)

    if min_costo == reemplazo:
        operaciones[:] = operaciones_reemplazo
        conteo_operaciones.update(conteo_operaciones_reemplazo)
        cadena_formada = cadena_reemplazo
    elif min_costo == insercion:
        operaciones[:] = operaciones_insercion
        conteo_operaciones.update(conteo_operaciones_insercion)
        cadena_formada = cadena_insercion
    elif min_costo == borrado:
        operaciones[:] = operaciones_borrado
        conteo_operaciones.update(conteo_operaciones_borrado)
        cadena_formada = cadena_borrado
    elif min_costo == kill:
        operaciones[:] = operaciones_kill
        conteo_operaciones.update(conteo_operaciones_kill)
        cadena_formada = cadena_kill

    return min_costo, cadena_formada

if __name__ == "__main__":
    cadena_original = input("Ingrese la cadena original: ")
    cadena_objetivo = input("Ingrese la cadena objetivo: ")
    costo, ops, conteo_ops, cadena_formada = algoritmoFuerzaBruta(cadena_original, cadena_objetivo)
    print("Costo Fuerza Bruta:", costo)
    print("Operaciones:", ops)
    print("Conteo de operaciones:", conteo_ops)
    print("Cadena formada:", cadena_formada)

File: interfaz_algoritmos/lib/algoritmos/problema1/algoritmoFuerzaBruta.dart
const int COST_ADVANCE = 1;
const int COST_DELETE = 2;
const int COST_INSERT = 2;
const int COST_REPLACE = 3;
const int COST_KILL = 1;

List algoritmoFuerzaBruta(String cadenaOriginal, String cadenaObjetivo) {
  List<String> operaciones = [];
  String cadenaFormada = "";
  Map<String, int> conteoOperaciones = {
    "advance": 0,
    "delete": 0,
    "insert": 0,
    "replace": 0,
    "kill": 0
  };

  var resultado = algoritmoFuerzaBrutaRecursiva(cadenaOriginal, cadenaObjetivo,
      0, 0, 0, operaciones, conteoOperaciones, cadenaFormada);

  return [
    resultado[0], // costo total
    resultado[1], // operaciones
    resultado[2], // conteo de operaciones
    resultado[3], // cadena formada
  ];
}

List algoritmoFuerzaBrutaRecursiva(
    String cadenaOriginal,
    String cadenaObjetivo,
    int i,
    int j,
    int costoTotal,
    List<String> operaciones,
    Map<String, int> conteoOperaciones,
    String cadenaFormada) {
  if (i == cadenaOriginal.length && j == cadenaObjetivo.length) {
    return [costoTotal, operaciones, conteoOperaciones, cadenaFormada];
  }

  int mejorCosto = double.infinity.toInt();
  List<String> mejorOperaciones = List.from(operaciones);
  Map<String, int> mejorConteo = Map.from(conteoOperaciones);
  String mejorCadena = cadenaFormada;

  // Opción 1: Advance
  if (i < cadenaOriginal.length &&
      j < cadenaObjetivo.length &&
      cadenaOriginal[i] == cadenaObjetivo[j]) {
    List<String> nuevasOperaciones = List.from(operaciones);
    Map<String, int> nuevoConteo = Map.from(conteoOperaciones);

    nuevasOperaciones.add("advance at position $i");
    nuevoConteo["advance"] = nuevoConteo["advance"]! + 1;

    var resultadoAvance = algoritmoFuerzaBrutaRecursiva(
        cadenaOriginal,
        cadenaObjetivo,
        i + 1,
        j + 1,
        costoTotal + COST_ADVANCE,
        nuevasOperaciones,
        nuevoConteo,
        cadenaFormada + cadenaOriginal[i]);

    if (resultadoAvance[0] < mejorCosto) {
      mejorCosto = resultadoAvance[0];
      mejorOperaciones = resultadoAvance[1];
      mejorConteo = resultadoAvance[2];
      mejorCadena = resultadoAvance[3];
    }
  }

  // Opción 2: Replace
  if (i < cadenaOriginal.length && j < cadenaObjetivo.length) {
    List<String> nuevasOperaciones = List.from(operaciones);
    Map<String, int> nuevoConteo = Map.from(conteoOperaciones);

    nuevasOperaciones.add("replace ${cadenaOriginal[i]} with ${cadenaObjetivo[j]} at position $i");
    nuevoConteo["replace"] = nuevoConteo["replace"]! + 1;

    var resultadoReemplazo = algoritmoFuerzaBrutaRecursiva(
        cadenaOriginal,
        cadenaObjetivo,
        i + 1,
        j + 1,
        costoTotal + COST_REPLACE,
        nuevasOperaciones,
        nuevoConteo,
        cadenaFormada + cadenaObjetivo[j]);

    if (resultadoReemplazo[0] < mejorCosto) {
      mejorCosto = resultadoReemplazo[0];
      mejorOperaciones = resultadoReemplazo[1];
      mejorConteo = resultadoReemplazo[2];
      mejorCadena = resultadoReemplazo[3];
    }
  }

  // Opción 3: Insert
  if (j < cadenaObjetivo.length) {
    List<String> nuevasOperaciones = List.from(operaciones);
    Map<String, int> nuevoConteo = Map.from(conteoOperaciones);

    nuevasOperaciones.add("insert ${cadenaObjetivo[j]} at position $i");
    nuevoConteo["insert"] = nuevoConteo["insert"]! + 1;

    var resultadoInsercion = algoritmoFuerzaBrutaRecursiva(
        cadenaOriginal,
        cadenaObjetivo,
        i,
        j + 1,
        costoTotal + COST_INSERT,
        nuevasOperaciones,
        nuevoConteo,
        cadenaFormada + cadenaObjetivo[j]);

    if (resultadoInsercion[0] < mejorCosto) {
      mejorCosto = resultadoInsercion[0];
      mejorOperaciones = resultadoInsercion[1];
      mejorConteo = resultadoInsercion[2];
      mejorCadena = resultadoInsercion[3];
    }
  }

  // Opción 4: Delete
  if (i < cadenaOriginal.length) {
    List<String> nuevasOperaciones = List.from(operaciones);
    Map<String, int> nuevoConteo = Map.from(conteoOperaciones);

    nuevasOperaciones.add("delete ${cadenaOriginal[i]} at position $i");
    nuevoConteo["delete"] = nuevoConteo["delete"]! + 1;

    var resultadoBorrado = algoritmoFuerzaBrutaRecursiva(
        cadenaOriginal,
        cadenaObjetivo,
        i + 1,
        j,
        costoTotal + COST_DELETE,
        nuevasOperaciones,
        nuevoConteo,
        cadenaFormada);

    if (resultadoBorrado[0] < mejorCosto) {
      mejorCosto = resultadoBorrado[0];
      mejorOperaciones = resultadoBorrado[1];
      mejorConteo = resultadoBorrado[2];
      mejorCadena = resultadoBorrado[3];
    }
  }

  // Opción 5: Kill
  if (i < cadenaOriginal.length && j == cadenaObjetivo.length) {
    List<String> nuevasOperaciones = List.from(operaciones);
    Map<String, int> nuevoConteo = Map.from(conteoOperaciones);

    nuevasOperaciones.add("kill from position $i");
    nuevoConteo["kill"] = nuevoConteo["kill"]! + 1;

    int costoKill = costoTotal + COST_KILL;

    if (costoKill < mejorCosto) {
      mejorCosto = costoKill;
      mejorOperaciones = nuevasOperaciones;
      mejorConteo = nuevoConteo;
      mejorCadena = cadenaFormada;
    }
  }

  return [mejorCosto, mejorOperaciones, mejorConteo, mejorCadena];
}
